FreeBSD compatibility

blodan opened the pull request:
After 1.2.0 tensorflow broke on FreeBSD, provided changes makes it build again.

Also, when nsync has merged this PR:
https://github.com/google/nsync/pull/3

The nsync dependency needs to be bumped.

Commit:
Add FreeBSD compatibility

Commit:
change back file permissions

tensorflow-jenkins commented:
Can one of the admins verify this patch?

gunan commented on tensorflow/core/platform/env.cc:
```diff
@@ -266,6 +270,13 @@ string Env::GetExecutablePath() {
+  size_t exe_path_size = PATH_MAX;
+
+  if (sysctl(mib, 4, exe_path, &exe_path_size, NULL, 0) != 0) {
+    // Not sure what to do if it fails?
```
well, what is the nature of this failure?
What exactly are we checking for in this if block?

blodan replied:
If sysctl cant resolve the path it returns a non zero value, but i can't find anything about a scenario where it would fail, and I cant see any of the other OS:s checking for a failure in their resolver so I just left it unhandled, should i set exe_path to an empty value if it fails perhaps?

Or should i perhaps do a 
> return errors::Aborted("could not resolve path ", exe_path)

gunan replied:
Having a failure like this looks more correct to me. Let's do that.

gunan commented on tensorflow/core/platform/env.cc:
```diff
@@ -293,6 +304,9 @@ bool Env::LocalTempFilename(string* filename) {
     int32 tid = static_cast<int32>(tid64);
     int32 pid = static_cast<int32>(getpid());
+#elif defined(__FreeBSD__)
+    int32 tid = static_cast<int32>((long) pthread_self());
```
do we need the extra cast to long here?
Cant we just rely on the else branch a few lines below?

blodan replied:
Hmm, if its not casted to long first, this compile error appears:
>error: static_cast from 'pthread_t' (aka 'pthread *') to 'int32' (aka 'int') is not allowed
>int32 tid = static_cast<int32>(pthread_self());

I'm no c coder so don't really know if there is a more elegant solution to this issue :)

gunan replied:
I see. Then let's add a comment stating this, then I am OK with this.

blodan commented:
@gunan any chance you could give me answers to the two questions so we could move this forward and get it merged? :)

gunan reviewed (changes requested):
I am very sorry, this PR completely fell off my radar. Thanks for reminding.

Commit:
- Bumping abseil-cpp to latest for FreeBSD compat
- Adding comment about error if not casted to long first
- Return error message if resolving path failed on FreeBSD

blodan commented:
@gunan should be good to go now :)

gunan commented on tensorflow/core/platform/env.cc:
```diff
@@ -266,6 +270,13 @@ string Env::GetExecutablePath() {
+  size_t exe_path_size = PATH_MAX;
+
+  if (sysctl(mib, 4, exe_path, &exe_path_size, NULL, 0) != 0) {
+    return "Resolving ", exe_path, " failed";
```
Looking at this again, this string can easily be interpreted as a path.
Let's leave a comment saying "Resolution of the path failed."
and return ""

Sorry for the change of mind here.

Commit:
Return empty string when path resolver fails

blodan commented:
@gunan i've now changed the resolver error to return an empty string instead as requested

gunan commented:
Jenkins, test this please.

yifeif commented:
Jenkins, test this please.
